src/vm/database/sqlite.rs

```diff
@@ -294,6 +325,126 @@ impl <'a> ContractDatabase <'a> {
         }
     }
 
+    fn load_token(&self, contract_name: &str, token_name: &str) -> Result<SqliteToken> {
+        let identifier =
+            self.query_row(
+                "SELECT token_identifier FROM tokens_table WHERE contract_name = ? AND token_name = ?",
+                &[contract_name, token_name],
+                |row| row.get(0))
+            .ok_or(UncheckedError::UndefinedAssetType(token_name.to_string()))?;
+
+        Ok(SqliteToken {
+            token_identifier: identifier })
+    }
+
+    fn load_asset(&self, contract_name: &str, asset_name: &str) -> Result<SqliteAsset> {
+        let (identifier, key_type) : (_, String) =
+            self.query_row(
+                "SELECT asset_identifier, key_type FROM assets_table WHERE contract_name = ? AND asset_name = ?",
+                &[contract_name, asset_name],
+                |row| (row.get(0), row.get(1)))
+            .ok_or(UncheckedError::UndefinedAssetType(asset_name.to_string()))?;
+
+        Ok(SqliteAsset {
+            asset_identifier: identifier,
+            key_type: TypeSignature::deserialize(&key_type) })
+    }
+
+    pub fn create_token(&mut self, contract_name: &str, token_name: &str) {
+        self.execute("INSERT INTO tokens_table (contract_name, token_name) VALUES (?, ?)",
+                     &[contract_name, token_name]);
+    }
+
+    pub fn create_asset(&mut self, contract_name: &str, asset_name: &str, key_type: &TypeSignature) {
+
+        self.execute("INSERT INTO assets_table (contract_name, asset_name, key_type) VALUES (?, ?, ?)",
+                     &[contract_name, asset_name, &key_type.serialize()]);
+    }
+
+    // Asset functions return error if no such token exists
+    pub fn get_token_balance(&mut self, contract_name: &str, token_name: &str, principal: &PrincipalData) -> Result<i128> {
+        let descriptor = self.load_token(contract_name, token_name)?;
+
+        let params: [&ToSql; 3] = [&TOKEN_TYPE,
+                                   &descriptor.token_identifier,
+                                   &principal.serialize()];
+        let sql_result: Option<i128> = 
+            self.query_row(
+                "SELECT value FROM data_table WHERE data_type = ? AND data_store_identifier = ? AND key = ? ORDER BY data_identifier DESC LIMIT 1",
+                &params,
+                |row| row.get(0));
+        match sql_result {
+            None => Ok(0),
```

Comment:
This seems to contradict the comment above -- if the query returns 0 rows, then this always return `Ok(0)`?

Reply:
It errors if you try to lookup the balance for a token which wasn't defined using `define-token`

Reply:
So, this branch is never taken?  If so, then should this branch `panic!` instead?

Reply:
No -- this branch will be taken if you try to lookup the token balance of a principal that hasn't received any tokens previously.

For example, anyone who hasn't received tokens yet would hit that path in the following:
```
(define-token foo-tokens)
(define-public (get-my-balance)
   (ok (get-balance foo-tokens tx-sender)))
```

The error case would be the following:
```
(define-token foo-tokens)
(define-public (get-my-balance)
   (ok (get-balance bar-tokens tx-sender)))
```

Because `bar-tokens` don't exist at all. Note, that the type checker would catch that. That's why it's an "UncheckedError" variant (poorly named error enum -- but it's the class of errors that _should_ be caught by the typechecker). These errors will only occur in testing, because in production, the typechecker will discover these errors before execution.